### DISCUSSION TITLE
Multiple GPUs utilisation ?

### DISCUSSION THREAD

**Discussion by nyue** [mitsuba-renderer/mitsuba3#281] [Q&A]
Hi,

  I have temporary access to a remote workstation that has multiple nvidia A100 GPUs

  I am wondering if Mitsuba3 will be able to utilise all the GPUs (in paralle) on such a machine ?

  Is there some options/flag I can set to build Mitsuba3 from source that can utilise multiple GPUs ?

Cheers

**Comment by Speierers** [comment]
Hi @nyue,

Although Dr.Jit was designed with the ability to run kernels on multiple GPUs, this has never been tested in practice on our side. 

Likewise, we never tried to render scenes on multiple GPUs in parallel using Mitsuba 3. It is on our roadmap to support this eventually but you might need to modify the current codebase in order to achieve this today. 

Unfortunately, I do not have any experience with multi-GPU kernels using OptiX so I won't be able to give you any guidance on how to achieve this. If you are going down this road, please share your findings with us on this thread. It will be very helpful once we get to work on supporting this in the future..

**Reply by merlinND** [reply]
An easy alternative, if your workload allows, is to run multiple instances of Mitsuba concurrently, each one using a single GPU (`CUDA_VISIBLE_DEVICES=...`).

**Reply by nyue** [reply]
@merlinND Running multiple instances of Mitsuba to render a single frame or each instance render one frame from an animated sequence ?

**Reply by merlinND** [reply]
I guess that's up to you, especially if you're using the Python API it's very flexible!

If you're doing standard forward rendering, you could render a single frame over multiple GPUs by allocating a part of the sample count to each GPU. Just make sure to render with different random seeds, and then average together the two resulting images.

Rendering separate frames is even easier, since no special handling is needed except for the environment variable `CUDA_VISIBLE_DEVICES=...`.

**Reply by nyue** [reply]
@Speierers I am a studio tools integration software engineer and do not have the requisite OptiX knowledge to contribute directly. However, as long as I have access to some multi-GPU workstation, I will continue to build Mitsuba3 and find ways to utilise all the GPUs and learn.
